contract/src/main/kotlin/com/r3/corda/sdk/token/contracts/FungibleTokenContract.kt

```diff
@@ -85,47 +84,46 @@ open class FungibleTokenContract : AbstractTokenContract<FungibleToken<TokenType
         require(hasZeroAmounts.not()) { "You cannot create output token amounts with a ZERO amount." }
         // There can be different owners in each move group. There map be one command for each of the signers publickey
         // or all the public keys might be listed within one command.
-        val inputOwningKeys: Set<PublicKey> = group.inputs.map { it.holder }.map { it.owningKey }.toSet()
-        val signers: Set<PublicKey> = moveCommands.flatMap { it.signers }.toSet()
+        val inputOwningKeys: Set<PublicKey> = group.inputs.map { it.holder.owningKey }.toSet()
+        val signers: Set<PublicKey> = moveCommands.flatMap(CommandWithParties<TokenCommand<T>>::signers).toSet()
         require(inputOwningKeys == signers) {
             "There are required signers missing or some of the specified signers are not required. A transaction " +
-                    "to move owned token amounts must be signed by ONLY ALL the owners of ALL the input owned " +
-                    "token amounts."
+                    "to move token amounts must be signed by ONLY ALL the owners of ALL the input token amounts."
         }
     }
 
     override fun handleRedeem(
-            group: InOutGroup<FungibleToken<TokenType>, IssuedTokenType<TokenType>>,
-            redeemCommand: CommandWithParties<TokenCommand<*>>
+            group: InOutGroup<FungibleToken<T>, IssuedTokenType<T>>,
+            redeemCommand: CommandWithParties<TokenCommand<T>>
     ) {
         val token = group.groupingKey
-        // There can be at most one output treated as a change paid back to the owner. Issuer is used to group states, so
-        // it will be the same as one for the input states.
-        require(group.outputs.size <= 1) { "When redeeming tokens, there must be zero or one output state." }
-        if (group.outputs.isNotEmpty()) {
-            require(group.outputs.single().amount > Amount.zero(token)) {
-                "In redeem groups there must hold that output token amount, if present, is greater than ZERO."
+        // There can be at most one output treated as a change paid back to the owner. Issuer is used to group states,
+        // so it will be the same as one for the input states.
+        group.outputs.apply {
+            require(size <= 1) { "When redeeming tokens, there must be zero or one output state." }
```

Comment:
The holder may want the change in multiple denominations; could we relax the requirement of maximum of one change output?

Reply:
We could do but the token selection algorithm doesn't support that for now. I'll raise it as a ticket for @roastario who is working on token selection.